salt/modules/aptpkg.py

```diff
@@ -158,6 +159,35 @@ def _check_apt():
         )
 
 
+def _is_locked(fd):
+    '''
+    Check of fd is locked
+    '''
+    try:
+        fcntl.lockf(fd, fcntl.LOCK_EX | fcntl.LOCK_NB)
+    except (IOError, OSError) as e:
+        return True
+    else:
+        fcntl.lockf(fd, fcntl.F_UNLCK)
+        return False
+
+
+def _wait_dpkg_lock():
+    '''
+    Wait for dpkg lock release
+    '''
+    if os.path.isfile("/var/lib/dpkg/lock-frontend"):
```

Comment:
This doesn't solve anything.

What if the lock happens after this check, but before the next statement?
According to some documentation, sometimes is /var/lib/dpkg/lock-frontend, sometimes /var/lib/dpkg/lock, depending on which tool is run to update/install stuff.

Reply:
Since 1 Apr, we are using this patch internally, and we don't have any error anymore. 
Maybe it's not the best fix but a don't know where we can fix that

Reply:
@Ch3LL https://github.com/saltstack/salt/blob/develop/tests/integration/states/test_pkg.py

Reply:
You can't avoid the race-condition, that's why I say you should handle the failing of the job by having a retry mechanism.

Reply:
- Error handling have to parse the apt error to catch the lock error, seem to me not a good idea.
- And retry on error, will retry for other error like missing package if the name has typo. 

But i understand our point of `if the lock happens after this check, but before the next statement` and i don't have answer for that.
When you use cloud instance `unattended-upgrades` is installed, and run on boot, saltstack failed 70% of the time

Reply:
> "Error handling have to parse the apt error to catch the lock error, seem to me not a good idea"

Can you explain why that is not a good idea. Is it not possible to parse for the lock file error?

Reply:
Error message can changed in future release and break the behaviour,

Put a can update my PR by catching `Could not get lock`
```
E: Could not get lock /var/lib/dpkg/lock-frontend - open (11: Resource temporarily unavailable)
E: Unable to acquire the dpkg frontend lock (/var/lib/dpkg/lock-frontend), is another process using it?
```
```python
def _call_apt(args, scope=True, **kwargs):
    '''
    Call apt* utilities.
    '''
    cmd = []
    if scope and salt.utils.systemd.has_scope(__context__) and __salt__['config.get']('systemd.scope', True):
        cmd.extend(['systemd-run', '--scope'])
    cmd.extend(args)

    params = {'output_loglevel': 'trace',
              'python_shell': False,
              'env': salt.utils.environment.get_module_environment(globals())}
    params.update(kwargs)

    _wait_dpkg_lock()
    return __salt__['cmd.run_all'](cmd, **params)
```

by somethings like that

```python
def _call_apt(args, scope=True, **kwargs):
    '''
    Call apt* utilities.
    '''
    cmd = []
    if scope and salt.utils.systemd.has_scope(__context__) and __salt__['config.get']('systemd.scope', True):
        cmd.extend(['systemd-run', '--scope'])
    cmd.extend(args)

    params = {'output_loglevel': 'trace',
              'python_shell': False,
              'env': salt.utils.environment.get_module_environment(globals())}
    params.update(kwargs)

    cmd_ret = __salt__['cmd.run_all'](cmd, **params)
    count = 0
    while "Could not get lock" in cmd_ret['stderr'] and count < 100:
        count += 1
        log.warning('Waiting for dpkg lock release: retrying... %s/100', count)
        time.sleep(3)
        cmd_ret = __salt__['cmd.run_all'](cmd, **params)
    return cmd_ret
```

Reply:
Again - your box seems to have acquired the dpkg lock from a graphical interface app, hence the lock-frontend. Can you check if you do a manual install like apt-get/apt/aptitude install foo from a terminal, if it's the same lock-frontend lock that's acquired? Some documentations mention '/var/lib/dpkg/lock'. When is when?

reference: https://askubuntu.com/questions/827212/which-is-the-process-using-apt-get-lock/827239
E: Could not get lock /var/lib/dpkg/lock - open (11: Resource temporarily unavailable)
E: Unable to lock the administration directory (/var/lib/dpkg/), is another process using it?
In theory the error messages can change, but in practice they rarely do, if ever. Salt is returning the underlying executable errors in almost all the cases (execution modules), so things will break pretty bad if that happens.

Doing this change in salt will give you an impression that your function is safe, and will wait, but it's not. I let other decide if it's meaningful for salt or not.

Reply:
Or if one day salt used python-apt we can handle the exception: 
https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/packaging/os/apt.py#L1193

I will change that by parsing the stderr.

Reply:
> Again - your box seems to have acquired the dpkg lock from a graphical interface app, hence the lock-frontend. Can you check if you do a manual install like apt-get/apt/aptitude install foo from a terminal, if it's the same lock-frontend lock that's acquired? Some documentations mention '/var/lib/dpkg/lock'. When is when?

`lock-frontend` is when you use apt, it's not link to graphical interface
`lock` is when you use dpkg directly